In [1]:
import socket
with socket.socket(socket.AF_INET, socket.SOCK_STREAM) as s:     # 소켓 s를 하나 열어서
    s.connect(('www.google.com', 80))          # www.google.com 이라는 호스트의 80번 포트에 접속합니다. 
    cmd = 'GET http://www.google.com HTTP/1.0\r\n\r\n'.encode() 
    s.send(cmd)                 # 접속이 되면 GET xxxxx 형식의 HTTP 프로토콜로 된 명령을 전송합니다. 
    data = s.recv(1024)    # 그리고 1024바이트만큼의 HTTP 응답을 수신합니다. 
print('Received', data)  # 어떤 응답을 수신했는지 볼까요?import requests
url = 'http://www.google.com'
response = requests.get(url)
response

Received b'HTTP/1.0 404 Not Found\r\nContent-Type: text/html; charset=UTF-8\r\nDate: Sat, 24 Oct 2020 06:07:01 GMT\r\nServer: gws\r\nContent-Length: 1561\r\nX-XSS-Protection: 0\r\nX-Frame-Options: SAMEORIGIN\r\n\r\n<!DOCTYPE html>\n<html lang=en>\n  <meta charset=utf-8>\n  <meta name=viewport content="initial-scale=1, minimum-scale=1, width=device-width">\n  <title>Error 404 (Not Found)!!1</title>\n  <style>\n    *{margin:0;padding:0}html,code{font:15px/22px arial,sans-serif}html{background:#fff;color:#222;padding:15px}body{margin:7% auto 0;max-width:390px;min-height:180px;padding:30px 0 15px}* > body{background:url(//www.google.com/images/errors/robot.png) 100% 5px no-repeat;padding-right:205px}p{margin:11px 0 22px;overflow:hidden}ins{color:#777;text-decoration:none}a img{border:0}@media screen and (max-width:772px){body{background:none;margin-top:0;max-width:none;padding-right:0}}#logo{background:url(//www.google.com/images/branding/googlelogo/1x/googlelogo_color_150x54dp.png) no-rep

#### HTTP의 메시지는 request/status line과 header 그리고 body로 이루어져있는데요,
#### 요청 내용을 담은 request line은 요청 메소드와 타겟주소를 포함합니다.

#### Request Method (요청 메소드)

 - GET: read, 정보를 요청
 - HEAD: 헤더 정보만 요청
 - POST: create, 정보를 생성, 변형 할때 사용
 - PUT: update, 정보를 수정
 - DELETE: delete, 정보를 삭제
 - OPTIONS : 사용가능한 메소드 확인

### Response Code (응답 코드)

- 1xx: 조건부 응답
- 2xx: 성공
- 3xx: 리다이렉션 완료
- 4xx: 요청 오류(클라이언트에 문제가 있다는 뜻!)
- 5xx: 서버 오류(서버에 문제가 있다는 뜻!)
#### 많이 사용하는 응답 코드

- 200: 서버가 클라이언트 요청을 성공적으로 처리, 웹 페이지에서는 페이지 요청이 정상적으로 완료되면 반환

- 404: 찾을 수 없음. 존재하지 않는 URL을 요청하여 페이지를 접속 할 수 없을 때 반환하는 응답 코드.

### 서비스 제공을 위한 API
    API를 서비스의 형태로 제공하는 회사에 대해 들어보셨나요? 예를 들어 Weather Underground는 날씨 데이터API를 판매합니다. 여기 몇가지 유명한 API를 소개 합니다.

    구글 API
    Google APIs Explorer | Google Developers  https://developers.google.com/apis-explorer

    카카오 API
    Kakao Developers  https://developers.kakao.com/

    트위터 API
    API reference index  https://developer.twitter.com/en/docs/api-reference-index

    네이버 파파고 API
    Naver Papago API  https://developers.naver.com/docs/papago/

In [2]:
# 깃허브 연결

'''
minho@DESKTOP-K2023AG MINGW64 ~
$ cd ./aiffel/
data_preprocess/ offline/

minho@DESKTOP-K2023AG MINGW64 ~
$ cd ./aiffel/offline/

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline
$ git init
Initialized empty Git repository in C:/Users/minho/aiffel/offline/.git/

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ vi README.md

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ ls
1024_z_score.ipynb  README.md

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ git remote add
usage: git remote add [<options>] <name> <url>

    -f, --fetch           fetch the remote branches
    --tags                import all tags and associated objects when fetching
                          or do not fetch any tag at all (--no-tags)
    -t, --track <branch>  branch(es) to track
    -m, --master <branch>
                          master branch
    --mirror[=(push|fetch)]
                          set up remote as a mirror to push to or fetch from


minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ git remote add origin https://github.com/mingikim1998/offline.git

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ git status
On branch master

No commits yet

Untracked files:
  (use "git add <file>..." to include in what will be committed)
        1024_z_score.ipynb
        README.md

nothing added to commit but untracked files present (use "git add" to track)

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ git add .
warning: LF will be replaced by CRLF in 1024_z_score.ipynb.
The file will have its original line endings in your working directory
warning: LF will be replaced by CRLF in README.md.
The file will have its original line endings in your working directory

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ git commit -m "initialize"
Author identity unknown

*** Please tell me who you are.

Run

  git config --global user.email "you@example.com"
  git config --global user.name "Your Name"

to set your account's default identity.
Omit --global to set the identity only in this repository.

fatal: unable to auto-detect email address (got 'minho@DESKTOP-K2023AG.(none)')

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ git config --global user.email "babel5165@gmail.com"

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ git config --global user.name "offline"

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ git commit -m "first"
[master (root-commit) a322070] first
 2 files changed, 2246 insertions(+)
 create mode 100644 1024_z_score.ipynb
 create mode 100644 README.md

minho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)
$ git push origin master
Enumerating objects: 4, done.
Counting objects: 100% (4/4), done.
Delta compression using up to 4 threads
Compressing objects: 100% (3/3), done.
Writing objects: 100% (4/4), 10.49 KiB | 1.75 MiB/s, done.
Total 4 (delta 0), reused 0 (delta 0), pack-reused 0
To https://github.com/mingikim1998/offline.git
 * [new branch]      master -> master
'''

'\nminho@DESKTOP-K2023AG MINGW64 ~\n$ cd ./aiffel/\ndata_preprocess/ offline/\n\nminho@DESKTOP-K2023AG MINGW64 ~\n$ cd ./aiffel/offline/\n\nminho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline\n$ git init\nInitialized empty Git repository in C:/Users/minho/aiffel/offline/.git/\n\nminho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)\n$ vi README.md\n\nminho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)\n$ ls\n1024_z_score.ipynb  README.md\n\nminho@DESKTOP-K2023AG MINGW64 ~/aiffel/offline (master)\n$ git remote add\nusage: git remote add [<options>] <name> <url>\n\n    -f, --fetch           fetch the remote branches\n    --tags                import all tags and associated objects when fetching\n                          or do not fetch any tag at all (--no-tags)\n    -t, --track <branch>  branch(es) to track\n    -m, --master <branch>\n                          master branch\n    --mirror[=(push|fetch)]\n                          set up remote as a mirror to push to or fetch from\n\n\n

In [3]:
# api key: dc761cbf7bf87616d3d6c9c0fe914d06

import requests
myapi = "af44a23b3fa31260fd92b236a65bffa3"

#lat: 위도, #lon:경도
def makeurl(lat, lng):
    url = "https://api.openweathermap.org/data/2.5/onecall?lat={}&lon={}&appid={}".format(lat, lng, myapi)
    return url

response = requests.get(makeurl(30,120))
weather = response.json()
weather

{'lat': 30,
 'lon': 120,
 'timezone': 'Asia/Shanghai',
 'timezone_offset': 28800,
 'current': {'dt': 1603521957,
  'sunrise': 1603490868,
  'sunset': 1603531218,
  'temp': 292.22,
  'feels_like': 288.86,
  'pressure': 1011,
  'humidity': 39,
  'dew_point': 277.96,
  'uvi': 6.51,
  'clouds': 35,
  'visibility': 10000,
  'wind_speed': 3.13,
  'wind_deg': 315,
  'wind_gust': 0,
  'weather': [{'id': 802,
    'main': 'Clouds',
    'description': 'scattered clouds',
    'icon': '03d'}]},
 'minutely': [{'dt': 1603521960, 'precipitation': 0},
  {'dt': 1603522020, 'precipitation': 0},
  {'dt': 1603522080, 'precipitation': 0},
  {'dt': 1603522140, 'precipitation': 0},
  {'dt': 1603522200, 'precipitation': 0},
  {'dt': 1603522260, 'precipitation': 0},
  {'dt': 1603522320, 'precipitation': 0},
  {'dt': 1603522380, 'precipitation': 0},
  {'dt': 1603522440, 'precipitation': 0},
  {'dt': 1603522500, 'precipitation': 0},
  {'dt': 1603522560, 'precipitation': 0},
  {'dt': 1603522620, 'precipitation': 0

In [4]:
import urllib
def download(url):
    return urllib.request.urlopen(url)

In [5]:
from urllib.error import URLError, HTTPError, ContentTooShortError

def download(url):
    try:
        html = urllib.request.urlopen(url)
    except (URLError, HTTPError, ContentTooShortError) as e:
        print('Download error', e.reason)
        html = None
    return html

download('https://www.google.com')

In [6]:
def download(url):
    try:
        html = urllib.request.urlopen(url).read()
    except (URLError, HTTPError, ContentTooShortError) as e:
        print('Download error', e.reason)
        html = None
    return html

In [7]:
def download(url):
    try:
        html = urllib.request.urlopen(url).read()
    except (URLError, HTTPError, ContentTooShortError) as e:
        print('Download error', e.reason)
        html = None
    return html

In [8]:
download('https://www.google.com')

b'<!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="ko"><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"><meta content="/images/branding/googleg/1x/googleg_standard_color_128dp.png" itemprop="image"><title>Google</title><script nonce="g7vLEaf08RaBrJl8T+JrOA==">(function(){window.google={kEI:\'UM-TX7fOBcb7-QbVw77gAg\',kEXPI:\'0,202162,1151585,5662,730,224,5104,207,2414,790,10,1226,364,1499,611,206,383,246,5,1354,648,652,2800,314,3,65,768,6,211,284,867,113,452,320,89,325,17,322,2250,84,7,312,1116857,1197726,557,328984,13677,4855,32692,16114,28684,9188,8384,1325,3534,1361,9291,3024,2819,1924,11028,1,1812,4020,978,7933,5295,2054,920,873,10622,7432,1571,5524,4518,2777,919,2277,8,85,4304,1279,2212,530,149,1103,840,518,1521,4258,109,203,1137,2,2669,2024,1776,520,1947,2219,3,100,328,1284,16,2927,2247,1812,1787,3227,1030,960,855,7,6068,6286,4455,641,7540,337,4928,108,3407,908,2,3555,2397,7468,3277,3,576,970,865,4625,148,3502,2488,2342,3982,1661

    파이썬 라이브러리 BeautifulSoup, Requests 사용하기

    파이썬은 웹 크롤링을 위해 표준라이브러리보다 더 편리한 라이브러리가 있습니다. 
    바로 BeautifulSoup과 Requests 라이브러리 인데요. 정말 많이 쓰입니다. pip을 이용해 설치할 수 있어요.

In [10]:
import requests
url = 'http://www.google.com'
response = requests.get(url)
response

#HTTP의 Response 관련 메소드에 2xx는 응답관련 메소드라고 했습니다. 200은 응답에 성공이란 뜻이에요.

# 응답에 성공하면, 이어서 get()메소드의 text 멤버변수에 접근 할 수 있는데요, 
# text는 HTML을 반환합니다. 다음은 예외처리를 한 코드에요

<Response [200]>

In [11]:
def download2(url):
    try:
        response = requests.get(url)
        html = response.text
    except requests.ConnectionError:
        print('Connection error')
        html = None
    return html

download2('https://www.google.com')

'<!doctype html><html itemscope="" itemtype="http://schema.org/WebPage" lang="ko"><head><meta content="text/html; charset=UTF-8" http-equiv="Content-Type"><meta content="/images/branding/googleg/1x/googleg_standard_color_128dp.png" itemprop="image"><title>Google</title><script nonce="+mNk09ibIc7uM4lsuEkUVw==">(function(){window.google={kEI:\'BNCTX-eiGtKUmAWL0bXoAw\',kEXPI:\'0,18168,183994,1151585,5662,731,223,5105,206,3204,10,1226,364,926,573,612,205,383,246,5,1129,225,648,653,2798,315,3,66,767,217,283,869,112,453,408,324,2590,84,7,312,193,1116664,1197753,529,328985,13677,4855,32692,16114,19397,9287,9188,8384,4858,1362,283,9008,3021,4746,11028,1,1812,4020,978,7931,5297,2054,622,298,873,4192,6430,14528,235,4282,2777,919,2277,8,570,2226,1593,1279,2212,241,289,149,1103,840,517,1522,4258,312,1137,2,2669,2025,2295,1947,2229,93,328,1284,17,2926,2246,1813,1787,3229,2843,7,6068,6286,4454,642,7877,4928,108,3407,908,2,940,2615,2398,6334,4410,3,346,230,970,865,4625,148,3027,475,2041,8,439,2342,39

In [12]:
import requests
from bs4 import BeautifulSoup
html = requests.get('http://www.google.com')
soup = BeautifulSoup(html.text, 'html.parser')

In [13]:
soup.html.body

<body bgcolor="#fff"><script nonce="QKXj9x3IgeotToed6OOeEw==">(function(){var src='/images/nav_logo229.png';var iesg=false;document.body.onload = function(){window.n && window.n();if (document.images){new Image().src=src;}
if (!iesg){document.f&&document.f.q.focus();document.gbqf&&document.gbqf.q.focus();}
}
})();</script><div id="mngb"><div id="gbar"><nobr><b class="gb1">검색</b> <a class="gb1" href="http://www.google.co.kr/imghp?hl=ko&amp;tab=wi">이미지</a> <a class="gb1" href="http://maps.google.co.kr/maps?hl=ko&amp;tab=wl">지도</a> <a class="gb1" href="https://play.google.com/?hl=ko&amp;tab=w8">Play</a> <a class="gb1" href="http://www.youtube.com/?gl=KR&amp;tab=w1">YouTube</a> <a class="gb1" href="https://news.google.com/?tab=wn">뉴스</a> <a class="gb1" href="https://mail.google.com/mail/?tab=wm">Gmail</a> <a class="gb1" href="https://drive.google.com/?tab=wo">드라이브</a> <a class="gb1" href="https://www.google.co.kr/intl/ko/about/products?tab=wh" style="text-decoration:none"><u>더보기</u> »</a><

In [14]:
soup = BeautifulSoup("<span>Wow it's so good!!</span>", 'html.parser')
soup.span

<span>Wow it's so good!!</span>

In [20]:
# ''', """ 는 여러줄의 문자열을 입력할 때 사용
html='''<title>Fundamental</title> 
         <body>
          <p id='programming'>python</p> 
          <p id='programming'>java</p> 
          <p id='algorithm'>algorithm</p> 
          <p id='fundamental'>math</p> 
          <p id='programming'>C++</p> 
          </body>'''
soup = BeautifulSoup(html, 'html.parser')
soup.findAll({'p'})

[<p id="programming">python</p>,
 <p id="programming">java</p>,
 <p id="algorithm">algorithm</p>,
 <p id="fundamental">math</p>,
 <p id="programming">C++</p>]

In [22]:
[i.text for i in soup.findAll(['p'])]

['python', 'java', 'algorithm', 'math', 'C++']

In [16]:
soup.findAll('p', id='programming')

[<p id="programming">python</p>,
 <p id="programming">java</p>,
 <p id="programming">C++</p>]

In [17]:
# https://www.crummy.com/software/BeautifulSoup/bs4/doc/
# 직관적

In [23]:
download('https://m.stock.naver.com/marketindex/index.nhn')

b'\n\n\n\n\n<!-- header -->\n\n\n\n\n\n\n\n\n<!doctype html>\n\n<html lang="ko">\n<head>\n    <meta charset="utf-8">\n    <meta name="viewport" content="width=device-width,initial-scale=1.0,maximum-scale=1.0,minimum-scale=1.0,user-scalable=no">\n\n    \n    \n        \n        \n            <meta property="og:url" content="http://m.stock.naver.com/marketindex/index.nhn?menu=exchange"/>\n            \n                \n                    <meta property="og:title" content="\xeb\x84\xa4\xec\x9d\xb4\xeb\xb2\x84 \xec\xa6\x9d\xea\xb6\x8c"/>\n                \n                \n            \n            \n                \n                    <meta property="og:description" content="\xea\xb4\x80\xec\x8b\xac\xec\xa2\x85\xeb\xaa\xa9\xec\x9d\x98 \xec\x8b\xa4\xec\x8b\x9c\xea\xb0\x84 \xec\xa3\xbc\xea\xb0\x80\xeb\xa5\xbc \xea\xb0\x80\xec\x9e\xa5 \xeb\xb9\xa0\xeb\xa5\xb4\xea\xb2\x8c \xed\x99\x95\xec\x9d\xb8\xed\x95\x98\xeb\x8a\x94 \xea\xb3\xb3"/>\n                \n                \n            \n 

In [31]:
import requests
from bs4 import BeautifulSoup
html2 = requests.get('https://m.stock.naver.com/marketindex/index.nhn')
soup = BeautifulSoup(html2.text, 'html.parser')

In [32]:
soup.html.body

<body class="marketindex">
<!-- u_skip -->
<div class="u_skip">
<a href="#content">본문 바로가기</a>
</div>
<header class="home_header" id="header" role="banner">
<!-- gnb -->
<div class="Ngnb">
<span class="Ngnb_logo"><a class="Nlogo_link" href="https://m.naver.com/"><span class="Nicon_logo">NAVER</span></a></span>
<div class="Ngnb_service">
<h1 class="Nservice_item"><a href="https://m.stock.naver.com/index.nhn"><span class="Nicon_service">증권</span></a></h1>
</div>
<div class="Ngnb_tool">
<a class="Ntool_button _btn_my" href="#" onclick="nclk(this, 'gnb.my', '', '');"><span class="Nicon_my">MY</span></a>
</div>
</div>
<div class="pop_mylist _my_layer">
</div>
<!-- 검색 -->
<!-- 모바일 -->
<div class="Nsearch">
<div class="Nsearch_inner">
<div class="Nsearch_box">
<div class="Nsearch_box_inner">
<form action="/searchItem.nhn" id="search" method="get" name="search" onsubmit="return delayed_submit(this)">
<span class="Nbox_text"><input class="Nbox_input_text" id="keyword_top" name="keyword_input" o

In [33]:
soup.findAll({'option'})

[<option data-unit="KRW" value="1">대한민국</option>,
 <option data-unit="ZAR" value="69.54">남아프리카</option>,
 <option data-unit="NOK" value="122.38">노르웨이</option>,
 <option data-unit="NZD" value="755.64">뉴질랜드</option>,
 <option data-unit="TWD" value="39.42">대만</option>,
 <option data-unit="DKK" value="179.67">덴마크</option>,
 <option data-unit="RUB" value="14.78">러시아</option>,
 <option data-unit="MYR" value="271.44">말레이지아</option>,
 <option data-unit="MXN" value="53.92">멕시코</option>,
 <option data-unit="MNT" value="0.4">몽골</option>,
 <option data-unit="USD" selected="true" value="1128.5">미국</option>,
 <option data-unit="BHD" value="2993.21">바레인</option>,
 <option data-unit="BDT" value="13.3">방글라데시</option>,
 <option data-unit="VND" value="0.0487">베트남</option>,
 <option data-unit="BRL" value="201.78">브라질</option>,
 <option data-unit="BND" value="831.77">브루나이</option>,
 <option data-unit="SAR" value="300.9">사우디</option>,
 <option data-unit="SEK" value="128.87">스웨덴</option>,
 <option data-unit=